### PR TITLE
Added special catch for std::exception in GTEST_TEST_NO_THROW_

### PR DESCRIPTION
**Pull request by Youw**
Re-open #2084 with fixes.

### CODE REVIEW

**Comment by gennadiycivil** [comment]
@Youw Thank you for this. 
Thank your for this contribution. Could you please elaborate and provide more information on exactly the issue / problems this is trying to address.
What is the before/after effect this would create and how is this beneficial?
The older PR has a lot of conversation back . and forth and is hard to follow.

**Comment by Youw** [comment]
_Why_:

`ASSERT/EXPECT_NO_THROW` is non-informative when it comes to investigating of "why a unit-test failed", when it failed.

_Example_:
```cpp
BusinessLogic logic;
ASSERT_NO_THROW(logic.perform());
```

_Before_:
```
Expected: logic.perform() doesn't throw an exception.
Actual: it throws.
```
It is unclear why the execution has failed, and manual investigation/debugging required.

_After_:
```
Expected: logic.perform() doesn't throw an exception.
Actual: it throws std::exception-derived exception with description: Startup error: configuration file /etc/project/main.conf is missing.
```
(The example is completely made up, but real-life cases sometime ridiculously close to this one.)

---
Some notes (case "before"):
Simplest way to get the exception message is to omit `ASSERT_NO_THROW` at all, and std::exception will be caught by GTest later, but:
1) The log message won't include failed statement and line;
2) Some developers omitting `ASSERT_NO_THROW` on purpose, to use 1) instead;
3) As a result of 2): no explicit highlight of "what is actually tested" by a particular test case;

---
Of course, not all projects follow "throw only std::exception-derived/std::runtime_error" rule, and this patch cannot cover all possible cases, but it greatly improve some part of it.

**Comment by darionesi** [comment]
Is there any way to know if something can be printed as part of the message? Like “does it have a stream operator<<?” ? That may help with other objects being thrown that are not deriving from stream but can be printed as additional information

**Comment by Youw** [comment]
There is no way to know what "that something" is.
Basically, what we have:
```cpp
try {
    // test check body
}
catch(const std::exception& e) {
     // print e.what()
}
catch(...) { // the problem in "..."
    // we know that there is an exception, but don't know of which type
    // maximum what we can get - std::current_exception(), but that is not helpful
}

**Comment by darionesi** [comment]
Can’t the detection pattern be used to check if there is an overload of operator<<? 
https://people.eecs.berkeley.edu/~brock/blog/detection_idiom.php 
https://www.bfilipek.com/2019/07/detect-overload-from-chars.html 
just asking

**Comment by Youw** [comment]
One can use detection pattern in case if have an object of unknown type, i.e. when using template function/class.

In our case we don't have an object at all.

File: googletest/include/gtest/internal/gtest-internal.h [hunk 1]
```diff
@@ -1337,17 +1346,24 @@ constexpr bool InstantiateTypedTestCase_P_IsDeprecated() { return true; }
 
 #define GTEST_TEST_NO_THROW_(statement, fail) \
   GTEST_AMBIGUOUS_ELSE_BLOCKER_ \
-  if (::testing::internal::AlwaysTrue()) { \
+  if (::testing::internal::TrueWithString gtest_msg = std::string()) { \
```
Comment (EricWF) [inline googletest/include/gtest/internal/gtest-internal.h]: Why can't this be declared on the next line?

File: googletest/include/gtest/internal/gtest-internal.h [hunk 2]
```diff
@@ -1337,17 +1346,24 @@ constexpr bool InstantiateTypedTestCase_P_IsDeprecated() { return true; }
...
     try { \
       GTEST_SUPPRESS_UNREACHABLE_CODE_WARNING_BELOW_(statement); \
     } \
+    catch (std::exception const& e) { \
+      gtest_msg.value = " std::exception-derived exception with description: "; \
+      gtest_msg.value += e.what(); \
+      gtest_msg.value += "."; \
```
Comment (EricWF) [inline googletest/include/gtest/internal/gtest-internal.h]: The value of `what()` is implementation defined. Why does adding a dot make sense?

File: googletest/include/gtest/internal/gtest-internal.h [hunk 3]
```diff
@@ -1337,17 +1346,24 @@ constexpr bool InstantiateTypedTestCase_P_IsDeprecated() { return true; }
...
+    } \
     catch (...) { \
+      gtest_msg.value = "."; \
       goto GTEST_CONCAT_TOKEN_(gtest_label_testnothrow_, __LINE__); \
     } \
   } else \
     GTEST_CONCAT_TOKEN_(gtest_label_testnothrow_, __LINE__): \
-      fail("Expected: " #statement " doesn't throw an exception.\n" \
-           "  Actual: it throws.")
+      fail(("Expected: " #statement " doesn't throw an exception.\n" \
```
Comment (EricWF) [inline googletest/include/gtest/internal/gtest-internal.h]: `fail` won't take a `std::string`?

**Comment by darionesi** [comment]
> In our case we don't have an object at all.

right! that makes my idea tricky to implement 🤣

File: googletest/include/gtest/internal/gtest-internal.h (hunk 1 above)
Comment (Youw) [reply googletest/include/gtest/internal/gtest-internal.h]: Same reason as in `GTEST_TEST_THROW_` - `gtest_msg` is used in both `if` and `else`

File: googletest/include/gtest/internal/gtest-internal.h (hunk 3 above)
Comment (Youw) [reply googletest/include/gtest/internal/gtest-internal.h]: No, it doesn't. It expands to 4-th parameter of `AssertHelper` c-tor: https://github.com/google/googletest/blob/90a443f9c2437ca8a682a1ac625eba64e1d74a8a/googletest/include/gtest/gtest.h#L1772

File: googletest/include/gtest/internal/gtest-internal.h (hunk 2 above)
Comment (Youw) [reply googletest/include/gtest/internal/gtest-internal.h]: To make sure we do have a dot at the end of the sentence, as we have in case of any other exception.

**Comment by Youw** [comment]
Do we have explicit change requests on this one?

File: googletest/include/gtest/internal/gtest-internal.h (hunk 2 above)
Comment (CJ-Johnson) [reply googletest/include/gtest/internal/gtest-internal.h]: Instead of adding a period at the end, are you able to change this such that the output is like the following?

```
gtest_msg.value = "`std::exception`-derived exception thrown with description \""
                                + e.what()
                                "\"";
```

File: googletest/include/gtest/internal/gtest-internal.h [hunk 4]
```diff
@@ -1337,17 +1346,24 @@ constexpr bool InstantiateTypedTestCase_P_IsDeprecated() { return true; }
...
       GTEST_SUPPRESS_UNREACHABLE_CODE_WARNING_BELOW_(statement); \
     } \
+    catch (std::exception const& e) { \
+      gtest_msg.value = " std::exception-derived exception with description: "; \
+      gtest_msg.value += e.what(); \
+      gtest_msg.value += "."; \
+      goto GTEST_CONCAT_TOKEN_(gtest_label_testnothrow_, __LINE__); \
+    } \
     catch (...) { \
+      gtest_msg.value = "."; \
```
Comment (CJ-Johnson) [inline googletest/include/gtest/internal/gtest-internal.h]: What is the purpose of this?

File: googletest/include/gtest/internal/gtest-internal.h (hunk 1 above)
Comment (CJ-Johnson) [reply googletest/include/gtest/internal/gtest-internal.h]: Can this be written with `::testing::internal::ConstCharPtr` instead of introducing a new type?

File: googletest/include/gtest/internal/gtest-internal.h (hunk 4 above)
Comment (Youw) [reply googletest/include/gtest/internal/gtest-internal.h]: To match previous behavior. See unit-tests

File: googletest/include/gtest/internal/gtest-internal.h (hunk 1 above)
Comment (Youw) [reply googletest/include/gtest/internal/gtest-internal.h]: It cannot: we need to own a copy of a string. ConstCharPtr is non-owning a wrapper over `const char*`

File: googletest/include/gtest/internal/gtest-internal.h (hunk 2 above)
Comment (CJ-Johnson) [reply googletest/include/gtest/internal/gtest-internal.h]: That works, for me! Add the quotes around the error message and then keep the dot as well :)

File: googletest/include/gtest/internal/gtest-internal.h (hunk 4 above)
Comment (Youw) [reply googletest/include/gtest/internal/gtest-internal.h]: updated to be more explicit

File: googletest/include/gtest/internal/gtest-internal.h (hunk 2 above)
Comment (Youw) [reply googletest/include/gtest/internal/gtest-internal.h]: updated

**Comment by CJ-Johnson** [comment]
@Youw I apologize for not making this change request earlier. Can you mark the `TrueWithString`'s constructors and conversion operator as explicit? I'll try to merge after that's done.

**Comment by Youw** [comment]
Sure, I will.

Except I'm not sure whether explicit `operator bool` would allow using `TrueWithString`'s variable declaration within `if` statement

I'll try and update, it that is possible.

**Comment by Youw** [comment]
Explicit c-tor -  done.

Explicit `operator bool` isn't possible in given use-case.
The very purpose of the class is to be able to declare it in `if` statement, _always (implicitly) evaluate to `true`_ and to be able to store a `string`, so it can be modified inside `if` and available inside `else`.
With C++17 this would simply be like:
```cpp
if (string msg; AlwaysTrue()) {
// set msg here
}
else {
// using msg here
}
```

**Comment by CJ-Johnson** [comment]
This appears to work https://godbolt.org/z/Cfwuzc

Can you help me understand why it isn't able to be explicit?

**Comment by CJ-Johnson** [comment]
Also it appears this change it out of date. It looks like similar behavior was already added.

https://github.com/google/googletest/blob/ba513d2c9525a7c986c115ed5d603f2cf17c6016/googletest/include/gtest/internal/gtest-internal.h#L822-L826

https://github.com/google/googletest/blob/ba513d2c9525a7c986c115ed5d603f2cf17c6016/googletest/include/gtest/internal/gtest-internal.h#L1269-L1282

**Comment by Youw** [comment]
1) hmm, I checked it locally on my PC, and it didn't compile, I'll double check: maybe a specific compiler, or a bit different implementation

2) the change you mentioned is in function GTEST_TEST_THROW_. This PR is about GTEST_TEST_NO_THROW_

**Comment by Youw** [comment]
> This appears to work https://godbolt.org/z/Cfwuzc

> hmm, I checked it locally on my PC, and it didn't compile, I'll double check: maybe a specific compiler, or a bit different implementation

I'm not sure why, but I couldn't reproduce the issue with `explicit operator bool` I had locally. Maybe I was confused by a different issue I had, which I probably fixed later.

As of now - updated accordingly.

**Comment by CJ-Johnson** [comment]
I will now try again to land this change.

**Comment by CJ-Johnson** [comment]
I'm performing a global test run. This may take up to 24 hours at which point I will be able to submit it.

**Comment by CJ-Johnson** [comment]
The test run has completed and some breakages were found. Essentially, some users depend on GCC headers that, under `-fno-exceptions`, expose something to the effect of:

```
#define catch(e) \
  if(0)
```

This causes compilation failure for your change. Inside the `std::exception` catch block, the body references an undeclared symbol `e`. I think the best solution at the moment is to conditionally add this catch block based on whether exceptions are enabled or disabled. Something like:

```
#if GTEST_HAS_EXCEPTIONS

#define GTEST_TESTNOTHROW_CATCH_STD_EXCEPTION_()  \
  catch (std::exception const& e) {  \
    gtest_msg.value = "it throws std::exception-derived exception with description: \"";  \
    gtest_msg.value += e.what();  \
    gtest_msg.value += "\".";  \
    goto GTEST_CONCAT_TOKEN_(gtest_label_testnothrow_, __LINE__);  \
  }

#else  // GTEST_HAS_EXCEPTIONS

#define GTEST_TESTNOTHROW_CATCH_STD_EXCEPTION_()

#endif  // GTEST_HAS_EXCEPTIONS
```

That way, when exceptions are disabled, `e` will not be present in source and will thus not fail to compile. Is this an acceptable change?

**Comment by Youw** [comment]
updated as suggested, but I've failed adding a test for this case, so left it as is

**Comment by CJ-Johnson** [comment]
Fantastic! Thank you :)

I just kicked off the global test run again. Fingers crossed!
